newt/mfg/decode.go

```diff
@@ -110,6 +112,40 @@ func decodeBoolDflt(kv map[string]interface{}, key string,
 	}
 }
 
+func decodeExtra(kv map[string]interface{},
+	key string) (map[string]interface{}, error) {
+
+	extra := kv[key]
+	if extra == nil {
+		return nil, nil
+	}
+
+	if _, ok := extra.(map[interface{}]interface{}); !ok {
+		return nil, util.FmtNewtError(
+			"invalid \"%s\" field: must be a map", key)
+	}
+
+	// If `extra` contains any maps, they were unmarshalled as
+	// `map[interface{}]interface{}`.  These need to be converted to
+	// `map[string]interface{}` to be JSON compatible.
+	var fixMap func(m map[string]interface{})
+	fixMap = func(m map[string]interface{}) {
+		for k, v := range m {
+			if _, ok := v.(map[interface{}]interface{}); ok {
+				sm := cast.ToStringMap(v)
+				fixMap(sm)
+
+				m[k] = sm
```

Comment:
When this adds a new `string` key from the cast, does the previous `interface{}` key remain in the map?

Reply:
This code is a bit confusing.  There are actually two maps involved:
```
m map[string]interface{}
v.(map[interface{}]interface{})
```

(maybe I could have chosen better variable names!)

The only map that gets modified here is `m` which has `string` keys.  The code just replaces the value associated with a particular string key.